### PR TITLE
Support for CALL procedures

### PR DESCRIPTION
**Pull request by systay**
This PR makes it possible to run `CALL` queries against vtgate. 

For now, this is only allowed if either specifically target specific shards, or if you are running on an unsharded keyspace.

- [x] Parsing CALL
- [x] vtgate/vttablet planning of CALL
- [x] Table should read multi results from MySQL server
~~- [x] Send multi results from tablet over grpc~~
~~- [ ] Send multi results from vtgate to client over mysql protocol~~
- [x] Write docs https://github.com/vitessio/website/pull/672
- [x] Check the state of the connection to MySQL from the tablet after running a sproc,  so no transactions are left open
~~- [ ] Make sure CALL works in both streaming and non-streaming modes~~ It will be handled in separate PR

### CODE REVIEW

File: go/vt/vtgate/planbuilder/call_proc.go [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+		keyspace, err := vschema.DefaultKeyspace()
+		if err != nil {
+			return nil, err
+		}
+		ks = keyspace.Name
+	} else {
+		ks = stmt.Name.Qualifier.String()
+	}
+
+	dest, keyspace, _, err := vschema.TargetDestination(ks)
```
Comment (harshit-gangal) [inline go/vt/vtgate/planbuilder/call_proc.go:39]: we can change this logic to 

```
ks = ""
if !stmt.Name.Qualifier.IsEmpty() {
   ks = stmt.Name.Qualifier.String()
}
dest, keyspace, _, err := vschema.TargetDestination(ks)
```

checking for default keyspace if qualifier is empty happens in TargetDestination method call

File: go/vt/vttablet/tabletserver/query_executor.go [hunk 2]
```diff
@@ -137,7 +137,7 @@ func (qre *QueryExecutor) Execute() (reply *sqltypes.Result, err error) {
 		return qr, nil
 	case planbuilder.PlanSelectLock:
 		return nil, vterrors.Errorf(vtrpcpb.Code_FAILED_PRECONDITION, "%s disallowed outside transaction", qre.plan.PlanID.String())
-	case planbuilder.PlanSet, planbuilder.PlanOtherRead, planbuilder.PlanOtherAdmin, planbuilder.PlanFlush:
+	case planbuilder.PlanSet, planbuilder.PlanOtherRead, planbuilder.PlanOtherAdmin, planbuilder.PlanFlush, planbuilder.PlanCallProc:
```
Comment (harshit-gangal) [inline go/vt/vttablet/tabletserver/query_executor.go:140]: this should also be added at Line 205 for executing in transaction. Needs to be validated the behaviour of stored procedure call inside a transaction.

File: go/test/endtoend/vtgate/unsharded/main_test.go [hunk 3]
```diff
@@ -163,6 +179,21 @@ func TestSelectIntoAndLoadFrom(t *testing.T) {
...
+	vtParams := mysql.ConnParams{
+		Host: "localhost",
+		Port: clusterInstance.VtgateMySQLPort,
+	}
+	conn, err := mysql.Connect(ctx, &vtParams)
+	require.NoError(t, err)
+	defer conn.Close()
+	defer exec(t, conn, `delete from t1`)
+	exec(t, conn, `insert into t1(c1, c2, c3, c4) values (300,100,300,'foo'),(301,101,301,'bar')`)
+	assertMatches(t, conn, `CALL GetAllT1()`, `[[INT64(300) INT64(100) INT64(300) VARCHAR("foo")] [INT64(301) INT64(101) INT64(301) VARCHAR("bar")]]`)
```
[the remark this answers is not in the document]
Comment (harshit-gangal) [reply go/test/endtoend/vtgate/unsharded/main_test.go:194]: Also needs to add test for `When the procedure returns, a client program can also obtain the number of rows affected for the final statement executed within the routine`

File: go/test/endtoend/cluster/vtctlclient_process.go [hunk 4]
```diff
@@ -57,9 +59,10 @@ func (vtctlclient *VtctlClientProcess) ApplySchemaWithOutput(Keyspace string, SQ
 }
 
 // ApplySchema applies SQL schema to the keyspace
-func (vtctlclient *VtctlClientProcess) ApplySchema(Keyspace string, SQL string) (err error) {
-	_, err = vtctlclient.ApplySchemaWithOutput(Keyspace, SQL, "")
-	return err
+func (vtctlclient *VtctlClientProcess) ApplySchema(Keyspace string, SQL string) error {
+	message, err := vtctlclient.ApplySchemaWithOutput(Keyspace, SQL, "")
+
+	return vterrors.Wrap(err, message)
```
[the remark this answers is not in the document]
Comment (systay) [reply go/test/endtoend/cluster/vtctlclient_process.go:65]: we were running into test errors but without the message, it was difficult to understand what was going wrong.

this is test code so I didn't think adding more output would be a problem

File: go/sqltypes/result.go [hunk 5]
```diff
@@ -225,3 +226,13 @@ func (result *Result) AppendResult(src *Result) {
...
 	return ToNamedResult(result)
 }
+
+// IsMoreResultsExists returns true if the status flag has SERVER_MORE_RESULTS_EXISTS set
+func (result *Result) IsMoreResultsExists() bool {
+	return result.StatusFlags&0x0008 == 0x0008
+}
+
+// IsInTranscation returns true if the status flag has SERVER_STATUS_IN_TRANS set
+func (result *Result) IsInTranscation() bool {
```
Comment (systay) [inline go/sqltypes/result.go:236]: typo: `IsInTranscation`

File: go/vt/vttablet/endtoend/call_test.go [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+BEGIN
+    start transaction;
+END;`, `create procedure proc_tx_commit()
+BEGIN
+    commit;
+END;`, `create procedure proc_tx_rollback()
+BEGIN
+    rollback;
+END;`}
+
```
Comment (systay) [inline go/vt/vttablet/endtoend/call_test.go:55]: we should add procedures with in and out params and test both that in params work and that out params fail with the expected error

File: go/vt/vttablet/endtoend/sequence_test.go [hunk 7]
```diff
@@ -59,6 +59,7 @@ func TestSequence(t *testing.T) {
 			sqltypes.NewInt64(13),
 			sqltypes.NewInt64(3),
 		}},
+		StatusFlags: 0x22,
```
Comment (systay) [inline go/vt/vttablet/endtoend/sequence_test.go:62]: don't we have consts for these values? would make it easier to read

File: go/vt/vttablet/tabletserver/connpool/dbconn.go [hunk 8]
```diff
@@ -172,6 +172,23 @@ func (dbc *DBConn) ExecOnce(ctx context.Context, query string, maxrows int, want
 	return dbc.execOnce(ctx, query, maxrows, wantfields)
 }
 
+// FetchNext returns the next result set.
+func (dbc *DBConn) FetchNext(ctx context.Context, maxrows int, wantfields bool) (*sqltypes.Result, error) {
+	// Check if the context is already past its deadline before
+	// trying to fetch the next result.
+	select {
+	case <-ctx.Done():
+		return nil, fmt.Errorf("%v before reading next result set", ctx.Err())
```
Comment (systay) [inline go/vt/vttablet/tabletserver/connpool/dbconn.go:181]: nit: `vterrors.Wrap`?

**Review by harshit-gangal — APPROVED** [review]

**Pull request merged**